File: 06_Algorithm/leetcode_141.go
/**
 * File              : leetcode_141.go
 * Author            : Andy
 * Date              : 27.07.2022
 * Last Modified Date: 27.07.2022
 * Last Modified By  : Andy
 */

/*
主体思路：如果有环，两个指针在跑的时候会出现快的追上慢的情况，
边界问题是：fast不能为空，且fast.Nextb也不能为空
*/

func hasCycle(head *ListNode) bool {
	var fast = head
	for fast != nil && fast.Next != nil {
		fast = fast.Next.Next
		head = head.Next
		if fast == head {
			return true
		}
	}
	return false
}

func main() {
	hasCycle(head)
}

File: Algorithm/leetcode_141.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# File              : leetcode_141.py
# Author            : Andy
# Date              : 27.07.2022
# Last Modified Date: 27.07.2022
# Last Modified By  : Andy

# 主体思路是：一直遍历，如果有环，那么跑得快的一定会追上跑得慢的
# 这里直接用head, fast就可以了，head走一步，fast走两步，边界条件
# 是fast, fast.next不为空，如果fast.next为空，fast.next.next会报错

class Solution:
    def hasCycle(self, head):
        fast = head 
        while fast and fast.next:
            head = head.next
            fast = fast.next.next
            if fast is head:
                return True
        return False
